user/acl/man/man3/acl_calc_mask.3
.\" Access Control Lists manual pages
.\"
.\" (C) 2002 Andreas Gruenbacher, <a.gruenbacher@computer.org>
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd March 23, 2002
.Dt ACL_CALC_MASK 3
.Os "Linux ACL"
.Sh NAME
.Nm acl_calc_mask
.Nd calculate the file group class mask
.Sh LIBRARY
Linux Access Control Lists library (libacl, \-lacl).
.Sh SYNOPSIS
.In sys/types.h
.In sys/acl.h
.Ft int
.Fn acl_calc_mask "acl_t *acl_p"
.Sh DESCRIPTION
The
.Fn acl_calc_mask
function calculates and sets the permissions associated with the ACL_MASK
ACL entry of the ACL referred to by
.Va acl_p .
The value of the new permissions is the union of the permissions
granted by all entries of tag type ACL_GROUP, ACL_GROUP_OBJ, or ACL_USER.
If the ACL referred to by
.Va acl_p
already contains an ACL_MASK entry, its permissions are overwritten;
if it does not contain an ACL_MASK entry, one is added.
.Pp
If the ACL referred to by
.Va acl_p
does not contain enough space for the new ACL entry, then additional working
storage may be allocated. If the working storage cannot be increased in the
current location, then it may be relocated and the previous working storage
is released and a pointer to the new working storage is returned via
.Va acl_p .
.Pp
The order of existing entries in the ACL is undefined after this function.
.Pp
Any existing ACL entry descriptors that refer to entries in the ACL continue to
refer to those entries. Any existing ACL pointers that refer to the ACL
referred to by
.Va acl_p
continue to refer to the ACL.
.\" <AG>
.\" Conflict between requirements:
.\" (a) ACL may be relocated,
.\" (b) all pointers remain valid.
.\" </AG>
.Sh RETURN VALUE
.Rv -std acl_calc_mask
.Sh ERRORS
If any of the following conditions occur, the
.Fn acl_calc_mask
function returns
.Li -1
and sets
.Va errno
to the corresponding value:
.Bl -tag -width Er
.It Bq Er EINVAL
The argument
.Va acl
is not a valid pointer to an ACL.
.It Bq Er ENOMEM
The
.Fn acl_calc_mask
function is unable to allocate the memory required for an ACL_MASK ACL entry.
.El
.Sh STANDARDS
IEEE Std 1003.1e draft 17 (\(lqPOSIX.1e\(rq, abandoned)
.Sh SEE ALSO
.Xr acl_check 3 ,
.Xr acl_get_entry 3 ,
.Xr acl_valid 3 ,
.Xr acl 5
.Sh AUTHOR
Derived from the FreeBSD manual pages written by
.An "Robert N M Watson" Aq rwatson@FreeBSD.org ,
and adapted for Linux by
.An "Andreas Gruenbacher" Aq a.gruenbacher@computer.org .


History
Ralink AP SDK - History

Version 3.4.1.0
===============

New Feature:

- [WebUI/NVRAM] add RT3092AP DLS GUI
- [WebUI] RVT GUI

App update:

- [Reset] adjust Default settings
- [Script] support GPL wifi driver insert/remove module

Fixed bug:

- [GPIO] Fix bug for GPIO=24~51
- [GPIOTOOL] add RALINK_GPIO_HAS_5124 for gpio24~gpio51
- [WebUI] STA site_survey error
- [WebUI] AP/STA display/Setting bugs
- [WebUI] WAPI/WPA_Supplicant Certificate upload
- [BusyBox] Sysklogd - avoid the read-only filesystem
- [Raether] MTU bug
- [Raether] Reorder the sequence of ei_close
- [Script] greenap.sh supports for MBSSID
- [Compile] sdk will compile goahead before mattrix, so we should modify goahead's Makefile to compile mattrix first
- [VPN] Multiple PPTP VPN pass through sessions are able to connect successfully for RT2880
- [NVRAM] rt3052 + i2s enabled => load default button's gpio=43

Version 3.4.0.0
===============

New Chip Support:

- [SDK] RT3883 support
- [SDK] RT3662 support
- [SDK] RT3350 support

New Feature:

- [Linux] Ralink Flow Classifier
- [Linux] Linux-based Watchdog driver (/dev/watchdog)
- [WebUI] WPS supplicant support
- [SDK] More 3G data card support
- [App] Video Flow Classification Command
- [App] Firewall-- Portscan, Syn flood  prevention. Snort Support.
- [App] iTune Server support
- [App] User space watchdog daemon
- [App] cURL
- [App] Expat- C library for parsing XML
- [App] mpstat
- [App] user space daemon supported 802.11f protocol
- [App] net-snmp 5.5
- [Web] Add WPS GUI for 2nd wireless interface
- [Web] Add WAPI GUI for 1st wireless interface
- [Web] Add WPA Supplicant GUI
- [Web] Add Wizard
- [Lib] Perl Compatible Regular Expressions
- [Lib] zlib-1.2.3
- [Lib] libpcap-1.0
- [Lib] libncurse-5.7

Peripheral Components:

- [I2S] audio codec WM8750 support for I2S TX/RX
- [PCM] Multi-SPI for PCM SLIC supported

Fixed bug:

- [PCM] assign next unmask channel every DMA done.
- [I2S] Change audio codec sampling rate to the nearest frequency settings.
- [WebUI] 3G PIN, APN, Password, Username, and Dial number support
- [WebUI] 3G auto-connect mode support
- [USB] OTG USB driver bug fix
- [NVRAM] fix commit length threshold
- [NVRAM] fix memory leak bug
- [App] IGMP Proxy issues fixed. (see History in user/igmpproxy/src/rt3052.c)


user/acl/libacl/acl_calc_mask.c
/*
  File: acl_calc_mask.c

  Copyright (C) 1999, 2000
  Andreas Gruenbacher, <a.gruenbacher@computer.org>

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU Library General Public
  License as published by the Free Software Foundation; either
  version 2 of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Library General Public License for more details.

  You should have received a copy of the GNU Library General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#include "libacl.h"


/* 23.4.2 */
int
acl_calc_mask(acl_t *acl_p)
{
	acl_obj *acl_obj_p;
	acl_entry_obj *entry_obj_p, *mask_obj_p = NULL;
	permset_t perm = ACL_PERM_NONE;
	if (!acl_p) {
		errno = EINVAL;
		return -1;
	}
	acl_obj_p = ext2int(acl, *acl_p);
	if (!acl_obj_p)
		return -1;
	FOREACH_ACL_ENTRY(entry_obj_p, acl_obj_p) {
		switch(entry_obj_p->etag) {
			case ACL_USER_OBJ:
			case ACL_OTHER:
				break;
			case ACL_MASK:
				mask_obj_p = entry_obj_p;
				break;
			case ACL_USER:
			case ACL_GROUP_OBJ:
			case ACL_GROUP:
				perm |= entry_obj_p->eperm.sperm;
				break;
			default:
				errno = EINVAL;
				return -1;
		}
	}
	if (mask_obj_p == NULL) {
		mask_obj_p = __acl_create_entry_obj(acl_obj_p);
		if (mask_obj_p == NULL)
			return -1;
		mask_obj_p->etag = ACL_MASK;
		__acl_reorder_entry_obj_p(mask_obj_p);
	}
	mask_obj_p->eperm.sperm = perm;
	return 0;
}

